### PR TITLE
BUG: Preserve Series/DataFrame subclasses through groupby operations

### LINKED ISSUE #28330: groupby and resample methods do not preserve subclassed data structures
#### Code sample

```python
class MyDataFrame(pd.DataFrame):
    @property
    def _constructor(self):
        return MyDataFrame

dates = pd.date_range('2019', freq='H', periods=1000)
my_df = MyDataFrame(np.arange(len(dates)), index=dates)

print(type(my_df)) 
# __main__.MyDataFrame (✓)

print(type(my_df.diff())) 
# __main__.MyDataFrame (✓)

print(type(my_df.sample(1))) 
# __main__.MyDataFrame (✓)

print(type(my_df.rolling('5H').mean()))
# __main__.MyDataFrame (✓)

print(type(my_df.groupby(my_df.index.dayofweek).mean()))
# pandas.core.frame.DataFrame (✘)

print(type(my_df.resample('D').mean()))
# pandas.core.frame.DataFrame (✘)

```
#### Problem description

Originally posted on [SO](https://stackoverflow.com/questions/57796464/pandas-groupby-resample-etc-for-subclassed-dataframe?noredirect=1#comment102088835_57796464).

The intended behaviour for chain-able methods on subclassed data structures is clearly that the operation returns an instance of the subclass (i.e. `MyDataFrame`), rather than the native type (i.e. `DataFrame`). This is the current behaviour for most operations (e.g., slicing, sampling, sorting) but not resample and groupby.

Currently groupby and resample both return explicitly constructed pandas datatypes, e.g. here:

https://github.com/pandas-dev/pandas/blob/ac693331400bce4747c63aa76c53c6e3488933ae/pandas/core/groupby/generic.py#L338

To get the expected behaviour, the intermediary classes (e.g. DataFrameGroupBy) would need to retain information about the calling class so that the appropriate constructor can be used (i.e. one of `_constructor` or `_constructor_sliced`  or `_constructor_expanddim`).

Note that operations that use Window and Rolling already appear have the expected behaviour because these assemble their results via a call to `concat` such as this one:

https://github.com/pandas-dev/pandas/blob/171c71611886aab8549a8620c5b0071a129ad685/pandas/core/window.py#L325

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 0.25.1
numpy            : 1.17.1
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.2.3
setuptools       : 40.8.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.8.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.1
numexpr          : 2.7.0
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : None
tables           : 3.5.2
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by JBGreisman**
This pull request fixes a bug that caused subclassed Series/DataFrames to revert to `pd.Series` and `pd.DataFrame` after `groupby()` operations. This is a follow-up on a previously abandoned pull request (#28573)

- [x] closes #28330
- [x] tests added / passed
- [x] passes black pandas
- [x] passes git diff upstream/master -u -- "*.py" | flake8 --diff
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.1.0.rst [hunk 1]
```diff
@@ -701,6 +701,7 @@ Groupby/resample/rolling
 - Bug in :meth:`DataFrame.groupby` where a ``ValueError`` would be raised when grouping by a categorical column with read-only categories and ``sort=False`` (:issue:`33410`)
 - Bug in :meth:`GroupBy.first` and :meth:`GroupBy.last` where None is not preserved in object dtype (:issue:`32800`)
 - Bug in :meth:`Rolling.min` and :meth:`Rolling.max`: Growing memory usage after multiple calls when using a fixed window (:issue:`30726`)
+- Bug where subclasses were not preserved through groupby operations (:issue:`28330`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:704]: can you enumerate which operations here

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -2047,7 +2051,13 @@ def ngroup(self, ascending: bool = True):
         """
         with _group_selection_context(self):
             index = self._selected_obj.index
-            result = Series(self.grouper.group_info[0], index)
+
+            # GH28330 preserve subclassed Series/DataFrames through calls
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:2055]: why is this needed?

File: pandas/core/groupby/groupby.py [hunk 3]
```diff
@@ -2109,7 +2119,12 @@ def cumcount(self, ascending: bool = True):
         with _group_selection_context(self):
             index = self._selected_obj.index
             cumcounts = self._cumcount_array(ascending=ascending)
-            return Series(cumcounts, index)
+
+            # GH28330 preserve subclassed Series/DataFrames through calls
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:2123]: why is this needed?

File: pandas/tests/groupby/test_groupby.py [hunk 4]
```diff
@@ -195,6 +195,42 @@ def f(grp):
     tm.assert_series_equal(result, e)
 
 
+@pytest.mark.parametrize(
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py:198]: make a new file test_groupby_subclass.

File: pandas/tests/groupby/test_groupby.py [hunk 5]
```diff
@@ -195,6 +195,42 @@ def f(grp):
...
+    "obj",
+    [
+        tm.SubclassedDataFrame({"A": np.arange(0, 10)}),
+        tm.SubclassedSeries(np.arange(0, 10), name="A"),
+    ],
+)
+def test_groupby_preserves_subclass(obj, groupby_func):
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py:205]: does this cover *all* of the changes? e.g. I see some that are in non-groupby files, please add tests for those as well.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.1.0.rst (hunk 1 above)
Comment (jreback) [reply doc/source/whatsnew/v1.1.0.rst:704]: you can also do general classes of things, e.g. fitler, transform, agg if it affects lots of functions.

File: pandas/tests/groupby/test_groupby.py (hunk 5 above)
Comment (JBGreisman) [reply pandas/tests/groupby/test_groupby.py:205]: To my knowledge, it does hit almost all of the changes in other files (I'll add any needed tests after combing through the codecov report).  For example:
- the changes in `frame.py` end up being hit by the `groupby.idxmin()` and `groupby.idxmax()` tests
- the changes to `concat.py` end up being hit by the `groupby.transform()` operations via `DataFrameGroupBy.__transform_general()`

I will add more targeted tests though to make sure other ways of calling those functions still preserve subclassing.

File: pandas/tests/groupby/test_groupby.py (hunk 4 above)
Comment (JBGreisman) [reply pandas/tests/groupby/test_groupby.py:198]: will do

File: pandas/core/groupby/groupby.py (hunk 2 above)
Comment (JBGreisman) [reply pandas/core/groupby/groupby.py:2055]: Returning the result of `GroupBy.ngroup()` as a `Series` ends up erasing the subclass information from `result._constructor_expanddim()`, which led to the default return of a `DataFrame`/`Series` instead of the subclass. 

It seemed to me that `self.obj` was either a `Series` or `DataFrame` instance (or subclass of those), depending on if was called by a `SeriesGroupBy` or `DataFrameGroupBy` so the if statement ensures that the proper `Series` (or subclass) is returned.

File: pandas/core/groupby/groupby.py (hunk 3 above)
Comment (JBGreisman) [reply pandas/core/groupby/groupby.py:2123]: This was the same logic as for `GroupBy.ngroup()`

**Comment by JBGreisman** [comment]
I updated my branch with additional tests, and a revised entry in the whatsnew rst file. 

Please let me know if there are any additional recommendations for how to proceed with this PR.

File: pandas/core/groupby/generic.py [hunk 6]
```diff
@@ -1101,7 +1108,8 @@ def _cython_agg_blocks(
                     deleted_items.append(locs)
                     continue
                 else:
-                    result = cast(DataFrame, result)
+                    # Cast to DataFrame-like type
+                    result = self.obj._constructor(result)
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1112]: this is not the same, is mypy complaining unless you do this?

File: pandas/core/groupby/generic.py [hunk 7]
```diff
@@ -1288,7 +1296,7 @@ def _wrap_applied_output(self, keys, values, not_indexed_same=False):
                         **kwargs, dtype_if_empty=object
                     )
                 else:
-                    backup = first_not_none._constructor(**kwargs)
+                    backup = first_not_none._constructor(**kwargs, dtype=object)
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1299]: really you had to add a dtype?

File: pandas/core/groupby/groupby.py (hunk 3 above)
Comment (jreback) [reply pandas/core/groupby/groupby.py:2123]: I would actually leave these. I don't think returning a subclass makes any sense here.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 7 above)
Comment (JBGreisman) [reply pandas/core/groupby/generic.py:1299]: I had added the dtype to avoid the DeprecationWarning when called with a `Series` (or subclass). You're correct that this is probably unnecessary/wrong -- I'll revert this back.

File: pandas/core/groupby/generic.py (hunk 6 above)
Comment (JBGreisman) [reply pandas/core/groupby/generic.py:1112]: yes -- mypy was complaining about this usage:
```python
result = cast(type(self.obj), result)
```
Please let me know if you have an alternative suggestion.

File: pandas/core/groupby/groupby.py (hunk 3 above)
Comment (JBGreisman) [reply pandas/core/groupby/groupby.py]: I reverted these changes -- `ngroup()` and `cumcount()` -- in the most recent set of commits.

Comment (JBGreisman) [reply pandas/core/groupby/groupby.py]: Thinking about this a bit more -- I'm not sure if it makes sense to have certain groupby-agg operations return the parent `Series` rather than a subclass. It seems more consistent to me to have all groupby.agg and groupby.transform preserve subclasses.

**Comment by JBGreisman** [comment]
Just to follow-up -- I have a few questions / comments on how best to proceed:

1. The change I made to appease mypy seems to pass all tests, but please let me know if you have a better way to handle that line.
2. Although I initially agreed with having `GroupBy.ngroup()` and `GroupBy.cumcount()` return `Series`, I think it makes sense to have them preserve subclasses (as I initially had it). This would make it so that all `groupby.agg()` / `groupby.transform()` operations return the calling subclass, rather than having two operations that return a different type. 
3. I have one failing build (Linux py37_np_dev), but inspecting the failure messages suggests that it's caused by the `DeprecationWarning` line that I had initially added a `dtype=object` to above. Please let me know if there's anything I should do about this line, or if I should leave it without the explicit dtype. 

Thanks for your help,
Jack

File: pandas/core/groupby/generic.py (hunk 7 above)
Comment (JBGreisman) [reply pandas/core/groupby/generic.py]: I changed the if statement above this line to check if `isinstance(first_not_none, Series)` so that it would also work for subclassed `Series`. This avoids the DeprecationWarnings that were being raised by my tests.

File: pandas/core/groupby/generic.py (hunk 6 above)
Comment (jreback) [reply pandas/core/groupby/generic.py:1082]: you can try
`assert isinstance(result, type(self.obj))`

cc @simonjayhawkins 

we don't want to actually call the constructor again

File: pandas/core/groupby/groupby.py [hunk 8]
```diff
@@ -1420,8 +1420,12 @@ def size(self):
         """
         result = self.grouper.size()
 
+        # GH28330 preserve subclassed Series/DataFrames through calls
         if isinstance(self.obj, Series):
+            result = self.obj._constructor(result)
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:1425]: try using assert here as well

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/groupby.py (hunk 8 above)
Comment (JBGreisman) [reply pandas/core/groupby/groupby.py:1425]: Thanks -- this sort of assertion worked to avoid unnecessary calls to the constructor

**Comment by JBGreisman** [comment]
Thanks for the mypy advice -- I managed to refactor that section a bit to only call `cast()` if needed and avoid calling the constructor again. 

The one additional change made here is to make sure `GroupBy.size()`, `GroupBy.ngroup()`, and `GroupBy.cumcount()` preserve subclassed types. While you had previously mentioned just keeping these return types as `Series`, I think these changes minimize the possibility for surprise when subclassing `DataFrame`/`Series` by ensuring that groupby reductions/transformations preserve the subclasses.

**Comment by JBGreisman** [comment]
@jreback -- please let me know if there are any additional changes to this PR that you think are necessary.

File: pandas/core/groupby/generic.py [hunk 9]
```diff
@@ -1071,7 +1079,10 @@ def _cython_agg_blocks(
                     deleted_items.append(locs)
                     continue
                 else:
-                    result = cast(DataFrame, result)
+                    # Ensure result is DataFrame-like type
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1082]: can you make this and elif (and include it rather than the else)

File: pandas/core/groupby/groupby.py [hunk 10]
```diff
@@ -2116,7 +2121,13 @@ def ngroup(self, ascending: bool = True):
         """
         with _group_selection_context(self):
             index = self._selected_obj.index
-            result = Series(self.grouper.group_info[0], index)
+
+            # GH28330 preserve subclassed Series/DataFrames
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:2125]: can you make this more concise e.g.

klass = self.obj_constructor if isinstance(self.obj, Series) else self.obj._constructor_sliced
result = klass(.....)

File: pandas/core/groupby/groupby.py [hunk 11]
```diff
@@ -2178,7 +2189,12 @@ def cumcount(self, ascending: bool = True):
         with _group_selection_context(self):
             index = self._selected_obj.index
             cumcounts = self._cumcount_array(ascending=ascending)
-            return Series(cumcounts, index)
+
+            # GH28330 preserve subclassed Series/DataFrames
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:2193]: same as above

File: pandas/core/groupby/groupby.py (hunk 10 above)
Comment (jreback) [reply pandas/core/groupby/groupby.py:2125]: also ok would be to add a property to GroupBy which return this e.g.

```
@property
def _constructor(self):
.....
```

**Review by jreback — CHANGES_REQUESTED** [review]
small comment, ping on green.

File: pandas/core/groupby/generic.py (hunk 9 above)
Comment (JBGreisman) [reply pandas/core/groupby/generic.py]: Looking back at this section, my changes were unnecessary. I reverted my changes in `DataFrameGroupBy._cython_agg_blocks()` back to the master branch -- this section is back to the original try-except-else statement.

File: pandas/core/groupby/groupby.py (hunk 10 above)
Comment (JBGreisman) [reply pandas/core/groupby/groupby.py]: Good idea -- I added a `GroupBy._constructor` property to consolidate the relevant logic in one place. This allowed me to change `ngroup`, `cumcount`, and `size` to use more concise statements.

**Comment by JBGreisman** [comment]
@jreback I made the recommend changes:
- I reverted the changes in `DataFrameGroupBy._cython_agg_blocks()` because they did not prove necessary to pass my tests.
- I added a `GroupBy._constructor` property to simplify the calls in `ngroup`, `cumcount`, and `size`.

File: pandas/core/groupby/generic.py [hunk 12]
```diff
@@ -1495,7 +1507,9 @@ def _transform_fast(self, result: DataFrame, func_nm: str) -> DataFrame:
                 res = maybe_cast_result(res, obj.iloc[:, i], how=func_nm)
             output.append(res)
 
-        return DataFrame._from_arrays(output, columns=result.columns, index=obj.index)
+        return type(self.obj)._from_arrays(
```
Comment (simonjayhawkins) [inline pandas/core/groupby/generic.py:1510]: why is this not self.obj._constructor like elsewhere?

**Review by simonjayhawkins — COMMENTED** [review]
@JBGreisman generally lgtm. one question.

File: pandas/core/groupby/generic.py (hunk 12 above)
Comment (JBGreisman) [reply pandas/core/groupby/generic.py]: Good catch -- I don't think there's a reason for this to be any different. I made the change to use `self.obj._constructor._from_arrays()` for consistency/clarity and it passes all the `groupby` tests.

**Comment by JBGreisman** [comment]
@simonjayhawkins Thanks for the comment -- I agree that line should have used `self.obj._constructor`, and changed it

File: pandas/core/groupby/groupby.py [hunk 13]
```diff
@@ -1182,6 +1182,16 @@ class GroupBy(_GroupBy[FrameOrSeries]):
     more
     """
 
+    @property
+    def _constructor(self) -> Type["Series"]:
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:1186]: hmm this should be FrameOrSeries I think.

why is there an else here?

Comment (JBGreisman) [reply pandas/core/groupby/groupby.py:1186]: I had thought this should be a Series because if `self.obj` is a DataFrame, it returns `self.obj._constructor_sliced`. Do you think I should name this property `_series_constructor` or something comparable to make that behavior more apparent?

The else was there because `mypy` was complaining about a missing return statement. I can restructure this with an assertion to avoid an else statement and keep `mypy` from complaining:
```python
@property
def _series_constructor(self) -> Type["Series"]:
    # GH28330 preserve subclassed Series/DataFrames                                                                                                                                                   
    if isinstance(self.obj, DataFrame):
        return self.obj._constructor_sliced
    assert isinstance(self.obj, Series)
    return self.obj._constructor
```
Please let me know if you have a better way to structure this.

Comment (jreback) [reply pandas/core/groupby/groupby.py:1186]: yes pls use an assertion

maybe @simonjayhawkins or @WillAyd can help with the annotation itself

Comment (JBGreisman) [reply pandas/core/groupby/groupby.py:1186]: thanks -- I changed `_constructor` to use the assertion as above. I think it would also make sense to change the name of the property to `_series_constructor` in order to clarify the return type, but I'll hold off for additional comments.

Comment (WillAyd) [reply pandas/core/groupby/groupby.py:1186]: I think the annotation is correct given the way it is implemented, though I am not sure about the implementation. Why do we need to dispatch to `constructor_sliced` for DataFrames? Seems slightly unnatural to have to force that

Comment (JBGreisman) [reply pandas/core/groupby/groupby.py:1186]: This ends up being used in `GroupBy.ngroup()`, `GroupBy.cumcount()`, and `GroupBy.size()`. Prior to these changes, these methods had returned `Series`, regardless of whether they were called from a `SeriesGroupBy` or `DataFrameGroupBy` object. I had updated this to still return a `Series`-type while preserving the subclasses -- to avoid things reverting back to `pd.Series` if they were called from a subclassed DataFrame/Series.

As such, the motivation for dispatching to `constructor_sliced` for DataFrames was to avoid changing the return type for these different `GroupBy` methods from their prior behavior. Do you think it would make sense to make a larger change here that alters these functions to have different return types if called from `SeriesGroupBy` vs. `DataFrameGroupBy`?

Comment (WillAyd) [reply pandas/core/groupby/groupby.py:1186]: Hmm that's unfortunate... Can you rename this property to `obj_constructor` instead? I find the current name a little confusing

Comment (WillAyd) [reply pandas/core/groupby/groupby.py:1186]: Or maybe even `_obj_1d_constructor` to be even more explicit. This is definitely for special cases so want to signal as such

Comment (JBGreisman) [reply pandas/core/groupby/groupby.py:1186]: yeah -- I think `_obj_1d_constructor` is most clear. I'll make the changes.

**Comment by JBGreisman** [comment]
As per the discussion above, the GroupBy property has been renamed to `_obj_1d_constructor` to clarify its expected return type, and I refactored it to avoid using an `else` statement.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @JBGreisman very nice, thanks for sticking with it and being responsive!

**Comment by JBGreisman** [comment]
no worries -- thanks for the help/suggestions!
